Sprint 4

vmishukov opened the pull request:
На ревью 4 спринт

vmishukov committed:
Финальная вёрстка макета

vmishukov committed:
Спринт 4/21: 4 → Тема 3/4: Реализация логики по макету → Урок 2/6
Добавил мок данные и действия для кнопок

vmishukov committed:
Спринт 4/21: 4 → Тема 3/4: Реализация логики по макету → Урок 3/6
Настроил вывод первого вопроса на экран

vmishukov committed:
Спринт 4/21: 4 → Тема 3/4: Реализация логики по макету → Урок 4/6
Добавил рамки при ответе на вопрос

vmishukov committed:
Реализовал логику перехода на следующий вопрос

vmishukov committed:
Алерт: показываем результат

vmishukov committed:
Поправил код

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -2,69 +2,168 @@ import UIKit
+                     , text: "Рейтинг этого фильма больше чем 6?"
+                     , corrcetAnswer: true),
+        QuizQuestion(image: "The Dark Knight",
+                     text: "Рейтинг этого фильма больше чем 6?",
```
Вопрос повторяется, рекомендую указать его как значение по умолчанию для поля text модели QuizQuestion

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -2,69 +2,168 @@ import UIKit
+    //массив вопросов
+    private let questions: [QuizQuestion] = [
+        QuizQuestion(image: "The Godfather"
+                     , text: "Рейтинг этого фильма больше чем 6?"
+                     , corrcetAnswer: true),
```
Запятые стоит перенести на предыдущие строчки

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -2,69 +2,168 @@ import UIKit
+    // константа с кнопкой для системного алерта
+  
     override func viewDidLoad() {
+        show(quiz: convert(model: questions[currentQuestionIndex]))
```
Настройку скругления картинки и толщины рамки лучше указать здесь, ее нужно указать лишь один раз:

        imageView.layer.masksToBounds = true // даём разрешение на рисование рамки
        imageView.layer.borderWidth = 8 // толщина рамки
        imageView.layer.cornerRadius = 20 // радиус скругления картинки

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -2,69 +2,168 @@ import UIKit
+        if isCorrect { // 1
+               correctAnswers += 1 // 2
+           }
+        imageView.layer.masksToBounds = true // даём разрешение на рисование рамки
+        imageView.layer.borderWidth = 1 // толщина рамки
+        imageView.layer.borderColor = isCorrect ? UIColor.ypGreen.cgColor : UIColor.ypRed.cgColor//
+        imageView.layer.cornerRadius = 6 // радиус скругления углов рамки
```
Здесь мы оставляем только присваивание цвета нашей рамке в зависимости от ответа, а все остально е убираем:

imageView.layer.borderColor = isCorrect ? UIColor.ypGreen.cgColor : UIColor.ypRed.cgColor

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -2,69 +2,168 @@ import UIKit
+        imageView.layer.borderColor = isCorrect ? UIColor.ypGreen.cgColor : UIColor.ypRed.cgColor//
+        imageView.layer.cornerRadius = 6 // радиус скругления углов рамки
+        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
+            self.showNextQuestionOrResults()
```
Перед показом результата еще нужно сделать рамку прозрачной:

imageView.layer.borderColor = UIColor.clear.cgColor

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -2,69 +2,168 @@ import UIKit
+  let buttonText: String
+}
 
+/*
```
Мок-данные можно удалить, они нам больше не понадобятся

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -2,69 +2,168 @@ import UIKit
+    // строка с названием фильма,
+    // совпадает с названием картинки афиши фильма в Assets
+    let image: String
+    // строка с вопросом о рейтинге фильма
+    let text: String
+    // булевое значение (true, false), правильный ответ на вопрос
+    let corrcetAnswer: Bool
+}
 
- Картинка: Tesla
- Настоящий рейтинг: 5,1
- Вопрос: Рейтинг этого фильма больше чем 6?
- Ответ: НЕТ
+struct QuizStepViewModel {
+  // картинка с афишей фильма с типом UIImage
+  let image: UIImage
+  // вопрос о рейтинге квиза
+  let question: String
+  // строка с порядковым номером этого вопроса (ex. "1/10")
+  let questionNumber: String
+}
 
+// для состояния "Результат квиза"
+struct QuizResultsViewModel {
+  // строка с заголовком алерта
+  let title: String
+  // строка с текстом о количестве набранных очков
+  let text: String
+  // текст для кнопки алерта
+  let buttonText: String
+}
```
Модели рекомендую написать в отдельном файле, знаю, что по ТЗ все должно быть написано в одном файле, но к моделям это не относится

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -2,69 +2,168 @@ import UIKit
         super.viewDidLoad()
     }
+    // метод вызывается, когда пользователь нажимает на кнопку "Нет"
+    @IBAction private func noButtonClicked(_ sender: Any) {
+        let currentQuestion = questions[currentQuestionIndex]
+        let givenAnswer = false
+        showAnswerResult(isCorrect: givenAnswer == currentQuestion.corrcetAnswer)
+    }
+    // метод вызывается, когда пользователь нажимает на кнопку "Да"
+    @IBAction private func yesButtonClicked(_ sender: Any) {
```
@IBAction принято писать в самом конце

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -2,69 +2,168 @@ import UIKit
+        let givenAnswer = true
+        showAnswerResult(isCorrect: givenAnswer == currentQuestion.corrcetAnswer)
+    }
+    @IBOutlet private weak var imageView: UIImageView!
+    @IBOutlet private weak var textLabel: UILabel!
+    @IBOutlet private weak var counterLabel: UILabel!
```
@IBOutlet принято писать в самом начале

vmishukov committed:
Отработка замечаний ревью:
- ❌ соблюдены расположение всех элементов, отступы, размеры кнопок, размеры рамки и скруглений
- ❌ изображения не деформируются, не обрезаются
- ❌ После выбора ответа на вопрос через 1 секунду автоматически появляется следующий вопрос и рамка ответа исчезает
И рекомендаций:
- 🍏В макете статус бар белый, а у тебя черный.
- 🍏Обычно используем такой порядок: сначала перечисляем свойства класса, затем функции. При этом также делим на приватные и публичные блоки. Функции перечисляют в таком порядке: initы, overrideы, блоки публичных и приватных функций, и далее блок c аннотацией @IBAction
- 🍏 Не забывай следить за отступами, общепринятые отступы помогают читать код быстрее. В общем случае поможет выделить весь файл cmd+A и зажать control+i
- 🍪Пользователь может случайно прожать два раза кнопку ответа или даже много и быстро понажимать на кнопки.
Это может привести к неверному увеличению счетчика, или лишнего вызова алерта с результатом.

protos17 commented on MovieQuiz/Presentation/Models.swift:
```diff
@@ -0,0 +1,35 @@
+    // строка с вопросом о рейтинге фильма
+    let text: String = "Рейтинг этого фильма больше чем 6?"
+    // булевое значение (true, false), правильный ответ на вопрос
+    let corrcetAnswer: Bool
```
Небольшая опечатка, рекомендую поправить на correctAnswer

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -2,71 +2,125 @@ import UIKit
+            correctAnswers += 1 // 2
+        }
+        imageView.layer.borderColor = isCorrect ? UIColor.ypGreen.cgColor : UIColor.ypRed.cgColor//
+        noButton.isEnabled = false
+        yesButton.isEnabled = false
```
Включение и выключение кнопок рекомендую реализовать в отдельном приватном методе, который в аргументах будет принимать булевое значение и исходя из нее будет блокировать и разблокировать кнопки
